### PR TITLE
Fix multipart/form-data requests support

### PR DESCRIPTION
**Pull request by diugalde**
Besides avoiding event.data to read the whole req.body when the content type is multipart/form-data, we need to configure the body parser (raw) to not generate a buffer with the payload.

### CODE REVIEW

**Comment by andresmgot** [comment]
Hi @diugalde,

I have built the NodeJS images with your changes, do you mind updating your branch so we can run the tests with the new image? The images are in `stable/nodejs/nodejs.jsonnet`. The new sha256 are:

 - NodeJS 6: `sha256:e72370621155aafe904e38e2c382dd3cd5e4058c7ad8baf9bda0d163c001399e`
 - NodeJS 8: `sha256:f4b210a7c45730fec888a8ff9ab853405536a361c2526306659c68487528892b`

Also, to ensure that the feature is working, do you mind adding a example as simple as possible in `stable/nodejs/examples/`? Then, to run that example as part of the tests, you need to add it to the NodeJS Makefile (`stable/nodejs/Makefile`). You can create two tasks `nodejs-multipart` and `nodejs-multipart-verify` and add it to the tasks `deploy` and `test` which are the ones executed in the CI. Something like:

```patch
diff --git a/stable/nodejs/Makefile b/stable/nodejs/Makefile
index 138c21d..5c659b4 100644
--- a/stable/nodejs/Makefile
+++ b/stable/nodejs/Makefile
@@ -16,8 +16,8 @@ build-all: build6 build8
 push-all: push6 push8
 
 # Testing jobs
-deploy: get-nodejs get-node-url-zip get-nodejs-custom-port get-nodejs-stream timeout-nodejs get-nodejs-deps get-nodejs-multi post-nodejs get-nodejs-extra-conf
-test: get-nodejs-verify get-node-url-zip-verify get-nodejs-custom-port-verify get-nodejs-stream-verify timeout-nodejs-verify get-nodejs-deps-verify get-nodejs-multi-verify post-nodejs-verify get-nodejs-extra-conf-verify
+deploy: get-nodejs get-node-url-zip get-nodejs-custom-port get-nodejs-stream timeout-nodejs get-nodejs-deps get-nodejs-multi post-nodejs get-nodejs-extra-conf nodejs-multipart
+test: get-nodejs-verify get-node-url-zip-verify get-nodejs-custom-port-verify get-nodejs-stream-verify timeout-nodejs-verify get-nodejs-deps-verify get-nodejs-multi-verify post-nodejs-verify get-nodejs-extra-conf-verify nodejs-multipart-verify
 
 get-node-url-zip:
        kubeless function deploy get-node-url-zip --runtime nodejs6 --handler index.helloGet --from-file https://github.com/kubeless/kubeless/blob/master/examples/nodejs/helloFunctions.zip?raw=true
@@ -96,3 +96,10 @@ get-nodejs-extra-conf-verify:
        kubectl rollout status deployment/nodejs-extra-conf && sleep 2
        $(eval POD := $(shell kubectl get po -l function=nodejs-extra-conf -o jsonpath="{.items[0].metadata.name}" 2>&1 || true))
        kubectl exec $(POD) -- npm config list | grep foo:bar
+
+nodejs-multipart:
+       kubeless function deploy nodejs-multipart --runtime nodejs6 --handler hello.handler --from-file examples/hello-multipart.js
+
+nodejs-multipart-verify:
+       kubeless function call nodejs-multipart | grep success
```

That way we can ensure the feature is working as expected and we make sure we don't break it in the future.

**Comment by diugalde** [comment]
Hello @andresmgot ,
I added a simple example and tried to modify the Makefile but I'm getting this error during the ci:
```
kubectl proxy -p 8085 &
curl \
  		--header "Content-Type:multipart/form-data" \
 		-F "prop=This is a normal prop" \
 		-F "file=@resources/input.txt" \
  		localhost:8085/api/v1/namespaces/default/services/hello-multipart:http-function-port/proxy/ |egrep hello.world
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (7) Failed to connect to localhost port 8085: Connection refused
make: *** [nodejs-multipart-verify] Error 1
make: Leaving directory `/home/circleci/project/stable/nodejs'
Exited with code 2
```
Locally, that's how I did it and it was working. I need to expose the function via http and use curl to perform a multipart request. Do you happen to know a workaround for this?

**Comment by andresmgot** [comment]
@diugalde you can try something like:

```
nodejs-multipart-verify:
	$(eval IP := $(shell minikube ip))
	$(eval TOKEN := $(shell kubectl get secrets -n kube-system  -o jsonpath={.items[0].data.token}))
	curl \
  		--header "Content-Type:multipart/form-data" \
  		--header "Authorization: Bearer $(TOKEN)" \
 		-F "prop=This is a normal prop" \
 		-F "file=@resources/input.txt" \
  		$(IP):8443/api/v1/namespaces/default/services/hello-multipart:http-function-port/proxy/ |egrep hello.world
```

That will be using the minikube IP and the default user to access the service. If the Makefile syntax gives you trouble you can always create a bash script and execute it in the make task.

**Comment by diugalde** [comment]
@andresmgot That seems correct. The ci build is now failing in a previous step (not sure why):
```
See also: go build, go get, go clean.
make[1]: *** [binary] Error 2
make[1]: Leaving directory `/home/circleci/.go_workspace/src/github.com/kubeless/kubeless'
mv: cannot stat ‘/home/circleci/.go_workspace/bin/kubeless’: No such file or directory
make: *** [kubeless] Error 1
Exited with code 2
```

File: stable/nodejs/Dockerfile.6 [hunk 1]
```diff
@@ -1,6 +1,6 @@
 FROM node:6
 
-RUN apt-get update &&  apt-get install git
+RUN apt-get update &&  apt-get install git && apt-get install curl
```
Comment (andresmgot) [inline stable/nodejs/Dockerfile.6]: you don't need to install curl in the NodeJS runtime, right?

File: stable/nodejs/examples/hello-multipart.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+const upload = multer({});
+
+// The request should send the file with this name.
+const FILENAME = 'file';
+const MULTIPART_TYPE = 'multipart/*';
+
+module.exports = {
+    handler: (event, context) => {
+        let fileContent;
+        if (event.extensions.request.is(MULTIPART_TYPE)) {
```
Comment (andresmgot) [inline stable/nodejs/examples/hello-multipart.js]: great example!

File: stable/nodejs/examples/package.json [hunk 3]
```diff
@@ -4,6 +4,7 @@
     "dependencies": {
         "end-of-stream": "^1.4.1",
         "from2": "^2.3.0",
-        "lodash": "^4.17.5"
+        "lodash": "^4.17.5",
```
Comment (andresmgot) [inline stable/nodejs/examples/package.json]: can you use a different file for the new example? something like `multipart-package.json`

File: stable/nodejs/resources/input.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+hello world
```
Comment (andresmgot) [inline stable/nodejs/resources/input.txt]: can you move this file within the `examples/` folder? (to avoid confusion)

**Comment by andresmgot** [comment]
The fix is already in master, please rebase to fix your issue

**Comment by andresmgot** [comment]
hi @diugalde any chance you can take a look to the required changes? I am preparing a new version of Kubeless and it would be good to include this feature in the NodeJS runtime.

**Comment by diugalde** [comment]
@andresmgot yeah! I'll work on that now so you can included it! Thanks!

**Comment by andresmgot** [comment]
@diugalde it's weird, I tried the make tasks in the CI and it says `Unauthorized` even after using an admin account. Changing the approach I made it work with the following task:

```
start-proxy:
        kubectl proxy& sleep 2

nodejs-multipart-verify: start-proxy
        curl -v \
  		--header "Content-Type:multipart/form-data" \
 		-F "prop=This is a normal prop" \
 		-F "file=@examples/resources/input.txt" \
                127.0.0.1:8001/api/v1/namespaces/default/services/hello-multipart:http-function-port/proxy/
```

But the function is returning an error:

```
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 8001 (#0)
> POST /api/v1/namespaces/default/services/hello-multipart:http-function-port/proxy/ HTTP/1.1
> Host: 127.0.0.1:8001
> User-Agent: curl/7.47.0
> Accept: */*
> Content-Length: 312
> Expect: 100-continue
> Content-Type:multipart/form-data; boundary=------------------------e708c902fc29838e
> 
< HTTP/1.1 100 Continue
< HTTP/1.1 200 OK
< Content-Length: 26
< Content-Type: text/plain; charset=utf-8
< Date: Fri, 28 Jun 2019 08:30:05 GMT
< X-Powered-By: Express
< 
* Connection #0 to host 127.0.0.1 left intact
You did not upload a file.
```

Do you know what could be happening? Are you able to make it work locally with the same image than in the PR?

**Comment by andresmgot** [comment]
I found the issue, the CI is not substituting the new images so it's not taking your changes. Let me fix that.

**Comment by andresmgot** [comment]
I have landed the fix, with the changes in master and my suggestion with the Makefile this should be ready to be landed.

**Comment by diugalde** [comment]
@andresmgot It's working now! Thank you!

**Comment by andresmgot** [comment]
great, thanks!

**Pull request merged**